lib/sqlalchemy/dialects/sqlite/base.py

```diff
@@ -402,6 +402,253 @@ def set_sqlite_pragma(dbapi_connection, connection_record):
+        constraint='id',
+        set_=dict(data='updated value')
+    )
+
+    conn.execute(do_update_stmt)
+
+Both methods supply the "target" of the conflict using either the
+named constraint or by column inference:
+
+* The :paramref:`.Insert.on_conflict_do_update.index_elements` argument
+  specifies a sequence containing string column names, :class:`_schema.Column`
+  objects, and/or SQL expression elements, which would identify a unique
+  index::
+
+
+    unique_index = schema.Index(
+        "unique_index_name",
+        my_table.c.id,
+        unique=True,
+    )
+
+    do_update_stmt = insert_stmt.on_conflict_do_update(
+        index_elements=['id'],
+        set_=dict(data='updated value')
+    )
+
+    do_update_stmt = insert_stmt.on_conflict_do_update(
+        index_elements=[my_table.c.id],
+        set_=dict(data='updated value')
+    )
+
+* When using :paramref:`.Insert.on_conflict_do_update.index_elements` to
+  infer an index, a partial index can be inferred by also specifying the
+  use the :paramref:`.Insert.on_conflict_do_update.index_where` parameter::
+
+    from sqlalchemy.dialects.sqlite import insert
+
+    unique_partial_index = schema.Index(
+        "unique_partial_index_name",
+        my_table.c.user_email,
+        unique=True,
+        sqlite_where=my_table.c.user_email.like('@gmail.com'),
+    )
+
+    stmt = insert(my_table).values(user_email='a@b.com', data='inserted data')
+    stmt = stmt.on_conflict_do_update(
+        index_elements=[my_table.c.user_email],
+        index_where=my_table.c.user_email.like('%@gmail.com'),
+        set_=dict(data=stmt.excluded.data)
+        )
+    conn.execute(stmt)
+
+* The :paramref:`.Insert.on_conflict_do_update.constraint` argument is
+  used to specify a constraint directly rather than inferring it.  This can
+  either be the name of the underlying columns of a UNIQUE constraint or
+  a PRIMARY KEY constraint. SQLite does not support specifying the names of
+  an UNIQUE or PRIMARY KEY constraint and INDEX within the ON CONFLICT Clause.
+
+  That is, the following will not work::
+    unique_constraint_pk = schema.PrimaryKeyConstraint(
+        my_table.c.id, name="uq_user_pk"
+    )
+
+    do_update_stmt = insert_stmt.on_conflict_do_update(
+        constraint='uq_user_pk',
+        set_=dict(data='updated value')
+    )
+
+ But the statements below will work::
+    do_update_stmt_A = insert_stmt.on_conflict_do_update(
+        constraint=unique_constraint_pk,
+        set_=dict(id=1, data='updated value')
+    )
+
+    do_update_stmt_B = insert_stmt.on_conflict_do_update(
+        constraint='id',
+        set_=dict(id=1, data='updated value')
+    )
+
+* The :paramref:`.Insert.on_conflict_do_update.constraint` argument may
+  also refer to a SQLAlchemy construct representing a constraint,
+  e.g. :class:`.UniqueConstraint` or :class:`.PrimaryKeyConstraint`.
+  In this use, only the name of the constraint's first column is used directly.
```

Comment:
**mike bayer** (zzzeek) wrote:

can the two bullets that talk about "constraint" just be one, shorter bullet that just shows the best way to use it?  particularly if we can point "constraint" just to the object itself.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2230